src/inspector_agent.cc

```diff
@@ -149,8 +154,165 @@ static int RegisterDebugSignalHandler() {
 
+class JsBindingsSessionDelegate : public InspectorSessionDelegate {
+ public:
+  JsBindingsSessionDelegate(Environment* env,
+                            Local<Object> session,
+                            Local<Object> receiver,
+                            Local<Function> callback)
+                            : env_(env),
+                              session_(env->isolate(), session),
+                              receiver_(env->isolate(), receiver),
+                              callback_(env->isolate(), callback) {
+    session_.SetWeak(this, JsBindingsSessionDelegate::Release,
+                     v8::WeakCallbackType::kParameter);
+  }
+
+  virtual ~JsBindingsSessionDelegate() {
+    session_.Reset();
+    receiver_.Reset();
+    callback_.Reset();
+  }
+
+  bool WaitForFrontendMessage() override {
+    return false;
+  }
+
+  void OnMessage(const v8_inspector::StringView& message) override {
+    Isolate* isolate = env_->isolate();
+    v8::HandleScope handle_scope(isolate);
+    Context::Scope context_scope(env_->context());
+    MaybeLocal<String> v8string =
+        String::NewFromTwoByte(isolate, message.characters16(),
+                               NewStringType::kNormal, message.length());
+    Local<Value> argument = v8string.ToLocalChecked().As<Value>();
+    TryCatch try_catch(isolate);
+    Local<Function> callback = callback_.Get(isolate);
+    Local<Object> receiver = receiver_.Get(isolate);
+    callback->Call(receiver, 1, &argument);
+    if (try_catch.HasCaught()) {
+      Local<String> string = try_catch.Message()->Get();
+      std::vector<char> buffer(1000, '\0');
+      size_t len = string->WriteUtf8(buffer.data(), buffer.size() - 1);
+      if (len > buffer.size() - 1) {
+        len = buffer.size() - 1;
+      }
+      buffer[len] = '\0';
+      fprintf(stderr, "Exception in the inspector message handler: %s\n",
+              buffer.data());
+    }
+  }
+
+  void Disconnect() {
+    Agent* agent = env_->inspector_agent();
+    if (agent->delegate() == this)
+      agent->Disconnect();
+  }
+
+ private:
+  static void Release(
+      const v8::WeakCallbackInfo<JsBindingsSessionDelegate>& info) {
+    info.SetSecondPassCallback(ReleaseSecondPass);
+    info.GetParameter()->session_.Reset();
+  }
+
+  static void ReleaseSecondPass(
+      const v8::WeakCallbackInfo<JsBindingsSessionDelegate>& info) {
+    JsBindingsSessionDelegate* delegate = info.GetParameter();
+    delegate->Disconnect();
+    delete delegate;
+  }
+  Environment* env_;
+  Persistent<Object> session_;
+  Persistent<Object> receiver_;
+  Persistent<Function> callback_;
+};
+
+Agent* GetInspectorChecked(Environment* env) {
+  Agent* agent = env->inspector_agent();
+  if (!agent) {
+    env->ThrowError("Inspector was not setup");
+    return nullptr;
+  }
+  return agent;
+}
+
+Local<v8::Private> DelegatePrivateKey(Isolate* isolate) {
+  return v8::Private::ForApi(
+      isolate, node::OneByteString(isolate, "Inspector#Delegate"));
```

Comment:
Can you use the [existing](https://github.com/nodejs/node/blob/9acd6c9aef399dbafa8a57399fbbb45376d7a81b/src/env.h#L66-L77) private symbol infrastructure?